From a26c7c763d0bf21c11356d6b64cd564406f18bc7 Mon Sep 17 00:00:00 2001
From: Sean Cross <xobs@kosagi.com>
Date: Thu, 23 Jul 2015 18:12:39 +0800
Subject: [PATCH 04/17] novena: drm: imx: HACK: Manually specify CRTCs

The CRTC assignment code doesn't realize that ldb needs two CRTCs that
must be a part of the same IPU when using both channels.

This patch hacks it so that the "IPU" only uses IPU1, and the DW-HDMI code
only uses IPU2.  It is Novena-specific and should not be used anywhere else.

Signed-off-by: Sean Cross <xobs@kosagi.com>
---
 drivers/gpu/drm/imx/dw_hdmi-imx.c  | 4 ++++
 drivers/gpu/drm/imx/imx-drm-core.c | 3 +++
 2 files changed, 7 insertions(+)

diff --git a/drivers/gpu/drm/imx/dw_hdmi-imx.c b/drivers/gpu/drm/imx/dw_hdmi-imx.c
index fe6becdcc29e..e9c7ff9e4c41 100644
--- a/drivers/gpu/drm/imx/dw_hdmi-imx.c
+++ b/drivers/gpu/drm/imx/dw_hdmi-imx.c
@@ -223,6 +223,10 @@ static int dw_hdmi_imx_bind(struct device *dev, struct device *master,
 	encoder = &hdmi->encoder;
 
 	encoder->possible_crtcs = drm_of_find_possible_crtcs(drm, dev->of_node);
+
+	/* Hack: Only use IPU2 */
+	encoder->possible_crtcs &= 0x0c;
+
 	/*
 	 * If we failed to find the CRTC(s) which this encoder is
 	 * supposed to be connected to, it's because the CRTC has
diff --git a/drivers/gpu/drm/imx/imx-drm-core.c b/drivers/gpu/drm/imx/imx-drm-core.c
index 5ea0c82f9957..ab37dcd2ce6a 100644
--- a/drivers/gpu/drm/imx/imx-drm-core.c
+++ b/drivers/gpu/drm/imx/imx-drm-core.c
@@ -151,6 +151,9 @@ int imx_drm_encoder_parse_of(struct drm_device *drm,
 
 	encoder->possible_crtcs = crtc_mask;
 
+	/* Hack: Only allow these devices to use IPU1 */
+	encoder->possible_crtcs &= 0x03;
+
 	/* FIXME: this is the mask of outputs which can clone this output. */
 	encoder->possible_clones = ~0;
 
-- 
2.20.1

